lib/main.js

```
@@ -0,0 +1,143 @@
+    var permissions = config.permission.permissions;
+    res.json(permissions);
+  });
+
+  /**
+   * Gets single role
+   */
+  app.get(
+    rolesBasePath + '/:id', ensureAuthenticated, function(req, res, next) {
+    var roleId = req.params.id;
+    if(config.permission.roleBaseUrl.length !== 0) {
+      roleId = config.permission.roleBaseUrl + '/' + roleId;
+    }
+    brPermission.getRole(req.user.identity, roleId, function(err, role) {
+      if(err) {
+        return next(err);
+      }
+      res.json(role);
+    });
+  });
+
+  /**
+   * Gets all existing roles
+   */
+  app.get(rolesBasePath, ensureAuthenticated, function(req, res, next) {
+    brPermission.getRoles(req.user.identity, function(err, roles) {
+      if(err) {
+        return next(err);
+      }
+      res.json(roles);
+    });
+  });
+
+  /**
+   * Inserts a newly created role to the database
+   */
+  app.post(
+    rolesBasePath, ensureAuthenticated, validate('services.role.postRole'),
+    function(req, res, next) {
+    brPermission.addRole(req.user.identity, req.body, function(err, result) {
+      if(err) {
+        if(database.isDuplicateError(err)) {
+          return next(new BedrockError(
+            'Duplicate role.', 'DuplicateRole', {
+            httpStatusCode: 409,
+            public: true
+          }));
+        }
+        return next(err);
+      }
+      res.status(201).json(result.role);
+    });
+  });
+
+  /**
+   * Edits a role that is already in the database
+   */
+  app.patch(
+    rolesBasePath + '/:id', ensureAuthenticated,
+    validate('services.role.patchRole'), function(req, res, next) {
+    // role ids in the database are not URLs
+    var updateOperations = req.body.filter(function(o) {
+      return o.op == 'updateRole';
+    }).map(function(o) {
+      return o.changes;
+    });
+    brPermission.updateRole(
+      req.user.identity, updateOperations[0], function(err, result) {
```

Calling this with `updateOperations[0]` is odd. It assumes there was only one `updateRole` operation. We should at least throw an error if there's more than one and that's not permitted.